citation/graphviz/data.py

```diff
@@ -103,6 +108,7 @@ def get_links(links_candidates, nodes_index):
 
 
 def get_nodes(nodes_candidates, filter_value, group_by):
+    # FIXME: the logic in this method is too complex / convoluted. simplify & refactor
```

Comment:
Can you clarify on what you expecting in this. I have modified to this below code.
def get_nodes(nodes_candidates, filter_value, group_by):
    # FIXME: the logic in this method is too complex / convoluted. simplify & refactor
    publications = Publication.api.primary(status="REVIEWED")
    nodes = []
    nodes_index = []
    for pub in nodes_candidates:
        publication = publications.get(pk=pub)
        if group_by == NetworkGroupBYType.SPONSOR.value:
            group_values = list(name[0] for name in publication.sponsors.all().values_list('name'))
        else:
            group_values = list(name[0] for name in publication.tags.all().values_list('name'))
        value = next((values for values in group_values if values in filter_value), None)
        group = "Others"
        if value:
            group = value
        nodes_index.append(pub)
    # FIXME: this particular expression needs to be cleaned up & refactored
        nodes.append({
            "name": pub, "group": group, 'tags': ', '.join(
            ['{0}'.format(s.name) for s in
                publication.tags.all()]), 'sponsors': ', '.join(
            ['{0}'.format(s.name) for s in
                publication.sponsors.all()]), "Authors": ', '.join(
            ['{0}, {1}.'.format(c.family_name, c.given_name_initial) for c in
                publication.creators.all()]), "title": publication.get(pk=pub).title
        })
    return nodes, nodes_index

Reply:
list comprehensions are great, but nested list comprehensions are difficult to parse. Be more explicit with your loops and what you are trying to do within each one and use temporary variables as necessary. This in turn can help facilitate discovery of better data structures / algorithms. In general, expect your first attempt at solving a problem to be naive and in need of further refinement so anything you can do to facilitate that future refinement will benefit you in the long run. Try to imagine modifications to your code that would extend it with additional functionality. What would you have to rewrite? How long would it take to for you to re-comprehend your own code after 6 months? A year? 5 years?

Simplicity > cleverness. But sometimes it takes a lot more work to find the simple solution than the clever one.